pg_analytics/src/hooks/query.rs

```diff
@@ -0,0 +1,30 @@
+use pgrx::*;
+use std::ffi::CStr;
+
+use crate::errors::ParadeError;
+
+pub trait Query {
+    // Extracts the query string from a PlannedStmt,
+    // accounting for multi-line queries where we only want a
+    // specific line of the entire query.
+    fn current_query_string(self, source_text: &CStr) -> Result<String, ParadeError>;
+}
+
+impl Query for *mut pg_sys::PlannedStmt {
```

Comment:
Happy to see `pg_analytics` adopting traits! But this might be a case where it's unnecessary.

You only really need traits for polymorphism, i.e. if a function doesn't want to know about the concrete type it might accept.

In this case, it looks like you're just trying to add a method to a `PlannedStmt`. You can just `impl *mut pg_sys::PlannedStmt` for that, there's no limit on how many `impl` blocks you can have, and they can be in any module.

As wonderful as traits are, unnecessarily defining them requires you to come up with another name, which can be problematic because of the "global" nature of traits. They can't easily be namespaced, so if there's another trait named `Query` in any one of our dependencies, we'll have a conflict.

Reply:
I had to define a trait because `pg_sys::PlannedStmt` is a primitive type. 

```rust
error[E0390]: cannot define inherent `impl` for primitive types
 --> pg_analytics/src/hooks/query.rs:6:1
  |
6 | impl *mut pg_sys::PlannedStmt {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: consider using an extension trait instead
```

Reply:
That makes sense, thanks for explaining